### DISCUSSION TITLE
Testing arrays and inner arrays without taking order into account

### DISCUSSION THREAD

**Discussion by galak75** [chaijs/chai#1522] [Support]
Hi everyone !

Based on this [excellent post](https://medium.com/building-ibotta/testing-arrays-and-objects-with-chai-js-4b372310fe6d), I tried to test arrays deep equality, making sure the item order is not relevant.

It works perfectly when having simple properties or object properties. But it fails when having an inner array property which order does not match. 

Here is a Typescript code example:
```
import { expect } from 'chai';

const chai = require('chai');
chai.use(require('deep-equal-in-any-order'));

describe('Test Chai members assertion', () => {
  it('test members assertion', () => {
    // fails as expected
    // expect([
    //   { name: 'John', age: 30 },
    //   { name: 'Bob', age: 25 },
    // ]).to.have.members([
    //   { name: 'Bob', age: 25 },
    //   { name: 'John', age: 30 },
    // ]);

    // succeeds as expected
    expect([
      { name: 'John', age: 30 },
      { name: 'Bob', age: 25 },
    ]).to.have.deep.members([
      { name: 'Bob', age: 25 },
      { name: 'John', age: 30 },
    ]);

    // fails as expected
    // expect([
    //   { name: 'John', age: 30 },
    //   { name: 'Bob', age: 25 },
    // ]).to.have.deep.ordered.members([
    //   { name: 'Bob', age: 25 },
    //   { name: 'John', age: 30 },
    // ]);

    // succeeds as expected
    expect([
      { name: 'John', age: 30, items: ['AA'] },
      { name: 'Bob', age: 25, items: ['BB', 'CC'] },
    ]).to.have.deep.members([
      { name: 'Bob', age: 25, items: ['BB', 'CC'] },
      { name: 'John', age: 30, items: ['AA'] },
    ]);

    // UNEXPECTEDLY FAILS!!!
    // I would like 'items' inner array order to be irrelevant (like main array order is)
    expect([
      { name: 'John', age: 30, items: ['AA'] },
      { name: 'Bob', age: 25, items: ['BB', 'CC'] },
    ]).to.have.deep.members([
      { name: 'Bob', age: 25, items: ['CC', 'BB'] },
      { name: 'John', age: 30, items: ['AA'] },
    ]);
  });
});
```

Do you have any idea how I could make sure inner arrays order is not relevant when comparing objects? (like the `items` inner property in the example above)

**Comment by galak75** [comment]
I kept digging for a solution, and I finally found this one, directly using the `deep-equal-in-any-order` chai plugin: 

```
    expect([
      { name: 'John', age: 30, items: ['AA'] },
      { name: 'Bob', age: 25, items: ['BB', 'CC'] },
    ]).to.deep.equalInAnyOrder([
      { name: 'Bob', age: 25, items: ['CC', 'BB'] },
      { name: 'John', age: 30, items: ['AA'] },
    ]);
```

But one question remains: is there a way to configure the way `members` assertion behaves? (e.g. by providing item matchers??)

**Reply by lll000111** [reply]
You can just `array.sort(myCustomCompareFn)`and deep-compare the sorted arrays. I would always use that over including yet another 3rd party package, no matter how small that is (you never know what will happen to those package and who might take control of them - if you only need a little bit of code, copy the relevant JS code from such packages directly and add comment-attributions inline).
